Split Login and Register, enable OIDC Registration.

uwwint opened the pull request:
This change introduces the ability to configure OIDC registration endpoints. In order to achieve this functionality Login and Register need to be split. The new functionality will do the following:

1. Add enduser registration endpoint to oidc_backends.xml
2. Expose them in /api/configuration 
3. Add a galaxy configuration option to disable local accounts
4. Change the frontend to:
    a. Split Login / Registration into two buttons
    <img width="1166" alt="Screenshot 2025-05-20 at 01 27 30" src="https://github.com/user-attachments/assets/7e912d8e-7621-4c81-9c8e-5440354777ea" />

    b. In case only one OIDC provider is configured and local accounts are disabled, Login will redirect the user to the OIDC endpoint.
    In case local accounts are enabled the behavior of previous galaxy is retained. 
    <img width="1166" alt="Screenshot 2025-05-20 at 01 27 39" src="https://github.com/user-attachments/assets/c7213df8-4aad-465c-adf9-d3f4ee4fb441" />
    c) In case local accounts are disabled and one OIDC provider with end user registration endpoint exists register will redirect the user straight to the registration of the OIDC provider. 
    In case no OIDC provider with end user registration is configured and local accounts are disabled, or local registration is disabled in galaxy.yml, registration is not shown on the UI (as per current behavior)
    If more than one OIDC provider with an endpoint are defined, or local accounts are enabled the Registration page will show all options. 
<img width="1166" alt="Screenshot 2025-05-20 at 01 27 44" src="https://github.com/user-attachments/assets/032d22ec-b9b2-45db-85ab-95651d7ba1cb" />

    

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  In order to test oidc registration, or disabling local accounts manual testing is still required. The best way is to test the following:
  1. disable local accounts, and configure one OIDC provider -> Only Login is visible on the UI, clicking login will directly point the user to the OIDC login.
  2. Enable local accounts and configure one OIDC provider -> Behavior as today. 
  3. Enable local accounts and configure one OIDC provider with end user registration endpoint -> Registration page shows button <Register with ...> 
<img width="1166" alt="Screenshot 2025-05-20 at 01 27 44" src="https://github.com/user-attachments/assets/4490b2a5-02e0-42d4-bf72-83c067e84e46" />
  3. Disable local accounts and configure one OIDC provider with end user registration endpoint -> Clicking registration will redirect the user to the end user registration point. 

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

uwwint committed:
Add Config opion disable_local_accounts

uwwint committed:
Rename allow_user_creation to allow_local_account_creation

uwwint committed:
add deprecation message for allow_user_creation on galaxy startup

uwwint committed:
add logic to disable local account creation when local accounts are disabled

uwwint committed:
Split Login and Registration. Add a new route /register/start duplicate the login page to not break test. This could be optimized further

uwwint committed:
If local account creation is disabled do not show username and password fields and login buttons

uwwint committed:
rebuild-config

uwwint committed:
add configuration option "end_user_registration_endpoint" to OIDC provider and push it tothe frontend

uwwint committed:
refactor ExternalLogin into helper and template implementation. Needs testing with Custos/CILogon

uwwint committed:
Login will now redirect in case there is only one OIDC provider and local accounts is disabled.

uwwint committed:
If local accounts are disabled and there is only one OIDC registration defined, push the user off to the configured registration.

uwwint committed:
Fix: if local accounts are enabled, the login page should be shown when clicking on login.

uwwint committed:
use config ref actual value

uwwint committed:
Seperate implementation of registration and login, fix all test cases, refctor code

uwwint committed:
add test cases for registration; refactoring

ahmedhamidawan committed:
refactoring clean up, fixes for various conditionals

ahmedhamidawan committed:
adjust `ExternalRegistration` jest for changes

ahmedhamidawan committed:
use `GLink` and `GButton` in login/registration; replacing `openurl` usage

ahmedhamidawan commented:
Thanks for these changes @uwwint !!!
I've made some minor styling/refactoring/linting/import ordering etc fixes, used Galaxy base components ...

Will try out all the use cases you've added next!

uwwint commented:
Hi @ahmedhamidawan do you need any help on setting up the Tests?

ahmedhamidawan commented:
> Hi @ahmedhamidawan do you need any help on setting up the Tests? 

Hey @uwwint 👋 !

Feel free to add/adjust tests, and make changes on top of mine if needed! 

Just make sure to lint the changes before you commit:
- `cd client && yarn run prettier --check`
- `cd client && yarn run eslint`
- `make format` (for backend)
- `tox -e mypy` (for backend)

ahmedhamidawan committed:
ensure horizontal or vertical separator not shown if local accounts disabled

Also some styling improvements to the buttons positioning.

ahmedhamidawan committed:
minute inline display fix

uwwint commented:
@marius-mather do you have capacity to give this a test - ? I can only get to it next week

marius-mather commented:
Hi @ahmedhamidawan , I've checked this PR out locally and I've confirmed that I'm able to disable local accounts, add a single OIDC provider,  and have the separate Login and Register buttons work as expected (going to the OIDC login, or the configured registration page). Everything's looking good to me

ahmedhamidawan committed:
fix selenium selectors for the now split login and register buttons

ahmedhamidawan committed:
remove `thin` prop from login/register `GLink`s to fix AXE violations

ahmedhamidawan committed:
change the new singular Login button's id to `user`

The Login button and logged-in user dropdown both serve as the primary user authentication area in the masthead. Using the same `#user` for both ensures consistent targeting in tours and tests, regardless of authentication state.

marius-mather commented:
Can this be merged now @nuwang ?

dannon commented on lib/galaxy/config/schemas/config_schema.yml:
```diff
@@ -2814,13 +2814,24 @@ mapping:
         type: bool
         default: true
         required: false
+        deprecated_alias: allow_user_creation
```
@ahmedhamidawan Thanks -- that's what I saw and assumed the alias hadn't been done, I should have checked but I was in a hurry to get the review out with all the pings.

dannon replied:
Yep
